### PR TITLE
[Core] Add support for async dynamic generators.

### PR DESCRIPTION
**Pull request by clarkzinzow**
This PR adds basic support for dynamic generators for async actors, i.e. [async generators](https://peps.python.org/pep-0525/) with a dynamic number of returns for async actors. This is mostly enabled by converting the async generator to a regular Python generator at task execution time, when processing the (yielded) task outputs in the main execution thread.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by stephanie-wang — APPROVED** [review]
Nice, this looks very clean!

Checked the code, and it seems like all the state used for generator returns is thread-local, so should be OK.

Could you also add a unit test for concurrent async generators before we merge?

**Comment by jjyao** [comment]
Please also update the doc :)

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -772,9 +774,15 @@ cdef void execute_task(
             function = execution_info.function
 
             if core_worker.current_actor_is_asyncio():
+                def predicate(obj):
+                    return (
+                        inspect.iscoroutinefunction(obj)
```
Comment (rkooo567) [inline python/ray/_raylet.pyx:779]: I've seen this predicate a lot. Maybe consider to convert it to a util function?

File: python/ray/util/tracing/tracing_helper.py [hunk 2]
```diff
@@ -514,6 +514,40 @@ async def _resume_span(
 
         return _resume_span
 
+    def async_gen_span_wrapper(method: Callable[..., Any]) -> Any:
```
Comment (rkooo567) [inline python/ray/util/tracing/tracing_helper.py:517]: do we test this from unit tests?

File: python/ray/tests/test_generators.py [hunk 3]
```diff
@@ -247,6 +292,26 @@ def generator(self, num_returns, store_in_plasma):
         g = Generator.remote()
         remote_generator_fn = g.generator
 
+    elif execution == "async_actors":
```
Comment (rkooo567) [inline python/ray/tests/test_generators.py:295]: Can we add a unit test if the execution is actually async? E.g., submit 10 of asyncio.sleep(1) and make sure the result is returned within 1.5 seconds or sth like that

File: python/ray/_private/async_compat.py [hunk 4]
```diff
@@ -45,6 +45,49 @@ async def wrapper(*args, **kwargs):
     return wrapper
 
 
+def async_gen_to_gen(async_gen, async_runner):
```
Comment (rkooo567) [inline python/ray/_private/async_compat.py:48]: Add types?

**Review by rkooo567 — COMMENTED** [review]
Mostly about testing comments

**Comment by rkooo567** [comment]
unstale

**Comment by rkooo567** [comment]
One of the core team members will take over this PR and fix tests.

**Comment by rkooo567** [comment]
This will replace this PR https://github.com/ray-project/ray/pull/34916

**Pull request closed without merging**